### PR TITLE
[JENKINS-29037] Allow for customization of the wait for condition step

### PR DESCRIPTION
**Pull request by mmitche**
Fixes an issue we were seeing where the step by default tends to run the contained closure very often initially.  When running lots of pipelines in parallel, this could cause problems if we had something like an http request within the body.  Allows for additional parameters to set the minimum recurrence period, the maximum recurrence period, and the units.

### CODE REVIEW

**Comment by jakauppila** [comment]
This would be great to have. We're seeing the same problem with the step as-is, we have some HTTP calls to check the status of another process that takes 60 min+, so we want to severely limit when and how often we're checking that.

File: src/main/java/org/jenkinsci/plugins/workflow/steps/WaitForConditionStep.java [hunk 1]
```diff
@@ -80,7 +116,6 @@
         }
 
         @Override public void onResume() {
-            recurrencePeriod = MIN_RECURRENCE_PERIOD;
```
Comment (jglick) [inline src/main/java/org/jenkinsci/plugins/workflow/steps/WaitForConditionStep.java]: You need to restore `currentRecurrencePeriod` to the minimum here; why was this deleted?

File: src/main/java/org/jenkinsci/plugins/workflow/steps/WaitForConditionStep.java [hunk 2]
```diff
@@ -56,16 +89,19 @@
          */
         private final String id = UUID.randomUUID().toString();
         private static final float RECURRENCE_PERIOD_BACKOFF = 1.2f;
-        static final long MIN_RECURRENCE_PERIOD = 250; // ¼s
-        static final long MAX_RECURRENCE_PERIOD = 15000; // ¼min
-        long recurrencePeriod = MIN_RECURRENCE_PERIOD;
-
-        Execution(StepContext context) {
+        long currentRecurrencePeriod;
```
Comment (jglick) [inline src/main/java/org/jenkinsci/plugins/workflow/steps/WaitForConditionStep.java]: The rename of the `recurrencePeriod` makes the rest of the diff harder to follow.

File: src/main/resources/org/jenkinsci/plugins/workflow/steps/WaitForConditionStep/config.jelly [hunk 3]
```diff
@@ -24,4 +24,15 @@ THE SOFTWARE.
 -->
 
 <?jelly escape-by-default='true'?>
-<j:jelly xmlns:j="jelly:core"/>
+<j:jelly xmlns:j="jelly:core" xmlns:f="/lib/form">
+    <f:entry field="minRecurrencePeriod" title="Minimum Recurrence Period">
```
Comment (jglick) [inline src/main/resources/org/jenkinsci/plugins/workflow/steps/WaitForConditionStep/config.jelly]: use `${%Some localizable text}` syntax

**Comment by jglick** [comment]
I think this purports to solve [JENKINS-29037](https://issues.jenkins-ci.org/browse/JENKINS-29037); marking accordingly.

**Comment by i386** [comment]
@mmitche hey there - are you still working on this one? is there any way we can help?

**Comment by mmitche** [comment]
@i386 I'll fix this up today.  I think I have the changes locally.

**Comment by mmitche** [comment]
Feedback addressed

**Comment by i386** [comment]
what say you @jglick ?

**Comment by i386** [comment]
@mmitche thanks mate!

File: src/main/java/org/jenkinsci/plugins/workflow/steps/WaitForConditionStep.java [hunk 4]
```diff
@@ -56,16 +89,23 @@
...
         private final String id = UUID.randomUUID().toString();
         private static final float RECURRENCE_PERIOD_BACKOFF = 1.2f;
-        static final long MIN_RECURRENCE_PERIOD = 250; // ¼s
-        static final long MAX_RECURRENCE_PERIOD = 15000; // ¼min
-        long recurrencePeriod = MIN_RECURRENCE_PERIOD;
-
-        Execution(StepContext context) {
+        long recurrencePeriod;
+        long maxRecurrencePeriod;
+        long minRecurrencePeriod;
```
Comment (jglick) [inline src/main/java/org/jenkinsci/plugins/workflow/steps/WaitForConditionStep.java]: When you are introducing non`transient` fields to a `Serializable` class you must define reasonable defaults for them in a `readResolve` method. In this case it means that a `waitUntil` block started prior to plugin upgrade will see zero values for `maxRecurrencePeriod` and `minRecurrencePeriod`, probably causing mayhem.

File: src/main/resources/org/jenkinsci/plugins/workflow/steps/WaitForConditionStep/config.jelly [hunk 5]
```diff
@@ -24,4 +24,15 @@ THE SOFTWARE.
...
-<j:jelly xmlns:j="jelly:core"/>
+<j:jelly xmlns:j="jelly:core" xmlns:f="/lib/form">
+    <f:entry field="minRecurrencePeriod" title="${%Minimum Recurrence Period}">
+        <f:number clazz="positive-number"/>
+    </f:entry>
+    <f:entry field="maxRecurrencePeriod" title="${%Maximum Recurrence Period}">
+        <f:number clazz="positive-number"/>
+    </f:entry>
+    <f:entry field="unit" title="Unit">
+        <f:select/>
```
Comment (jglick) [inline src/main/resources/org/jenkinsci/plugins/workflow/steps/WaitForConditionStep/config.jelly]: Should set a `default`.

File: src/main/resources/org/jenkinsci/plugins/workflow/steps/WaitForConditionStep/config.jelly [hunk 6]
```diff
@@ -24,4 +24,15 @@ THE SOFTWARE.
 -->
 
 <?jelly escape-by-default='true'?>
-<j:jelly xmlns:j="jelly:core"/>
+<j:jelly xmlns:j="jelly:core" xmlns:f="/lib/form">
+    <f:entry field="minRecurrencePeriod" title="${%Minimum Recurrence Period}">
+        <f:number clazz="positive-number"/>
+    </f:entry>
+    <f:entry field="maxRecurrencePeriod" title="${%Maximum Recurrence Period}">
+        <f:number clazz="positive-number"/>
```
Comment (jglick) [inline src/main/resources/org/jenkinsci/plugins/workflow/steps/WaitForConditionStep/config.jelly]: You need a `default` here, which would most easily be a constant in `DescriptorImpl`, moved there from the `Step` itself. [Documentation](https://github.com/jenkinsci/pipeline-plugin/blob/master/DEVGUIDE.md#handling-default-values)

**Comment by mmitche** [comment]
@jglick Responded to feedback.  I only added the min/max to the readResolve since recurrencePeriod will be set on resume.

**Comment by mmitche** [comment]
@jglick Anything else here?

**Comment by mmitche** [comment]
@abayer Thanks

**Comment by mmitche** [comment]
Anything else or can this be merged?

**Comment by szpak** [comment]
@abayer Have you been able to merge it? It seems still to be "in review".

File: src/main/java/org/jenkinsci/plugins/workflow/steps/WaitForConditionStep.java [hunk 7]
```diff
@@ -56,16 +91,29 @@
...
+        
+        private Object readResolve() {
+            minRecurrencePeriod = DescriptorImpl.DEFAULT_MIN_RECURRENCE_PERIOD;
+            maxRecurrencePeriod = DescriptorImpl.DEFAULT_MAX_RECURRENCE_PERIOD;
+            return this;
         }
 
         @Override public boolean start() throws Exception {
             body = getContext().newBodyInvoker().withCallback(new Callback(id)).start();
+            this.recurrencePeriod = step.unit.toMillis(step.getMinRecurrencePeriod());
```
Comment (dwnusbaum) [inline src/main/java/org/jenkinsci/plugins/workflow/steps/WaitForConditionStep.java]: Would `step.unit` be null here for running builds if we upgraded the plugin and then restarted  Jenkins?
